DigitalLearningSolutions.Web.Tests/Controllers/Login/LoginControllerTests.cs

```diff
@@ -92,12 +88,68 @@ public void Successful_sign_in_should_render_home_page()
                 .WithControllerName("Home").WithActionName("Index");
         }
 
+        [Test]
+        public void Successful_sign_in_with_local_return_url_should_redirect_to_return_url()
+        {
+            // Given
+            var expectedAdmin = UserTestHelper.GetDefaultAdminUser();
+            var expectedDelegates = new List<DelegateUser> { UserTestHelper.GetDefaultDelegateUser() };
+            GivenSignInIsSuccessful(expectedAdmin, expectedDelegates);
+
+            var returnUrl = "/some/other/page";
+            var urlHelper = controller.Url;
+            A.CallTo(() => urlHelper.IsLocalUrl(returnUrl)).Returns(true);
+
+            // When
+            var loginViewModel = LoginTestHelper.GetDefaultLoginViewModel();
+            loginViewModel.ReturnUrl = returnUrl;
+            var result = controller.Index(loginViewModel);
+
+            // Then
+            result.Should().BeRedirectResult().WithUrl(returnUrl);
+        }
+
+        [Test]
+        public void Successful_sign_in_with_nonlocal_return_url_should_render_home_page()
+        {
+            // Given
+            var expectedAdmin = UserTestHelper.GetDefaultAdminUser();
+            var expectedDelegates = new List<DelegateUser> { UserTestHelper.GetDefaultDelegateUser() };
+            GivenSignInIsSuccessful(expectedAdmin, expectedDelegates);
+
+            var returnUrl = "www.suspicious.com";
+            var urlHelper = controller.Url;
+            A.CallTo(() => urlHelper.IsLocalUrl(returnUrl)).Returns(false);
+
+            // When
+            var loginViewModel = LoginTestHelper.GetDefaultLoginViewModel();
+            loginViewModel.ReturnUrl = returnUrl;
+            var result = controller.Index(loginViewModel);
+
+            // Then
+            result.Should().BeRedirectToActionResult()
+                .WithControllerName("Home").WithActionName("Index");
+        }
+
         [Test]
         public void Successful_sign_in_should_call_SignInAsync()
         {
-            //Given
+            // Given
             var expectedAdmin = UserTestHelper.GetDefaultAdminUser();
             var expectedDelegates = new List<DelegateUser> { UserTestHelper.GetDefaultDelegateUser() };
+            GivenSignInIsSuccessful(expectedAdmin, expectedDelegates);
+
+            // When
+            controller.Index(LoginTestHelper.GetDefaultLoginViewModel());
+
+            // Then
+            A.CallTo(() => authenticationService.SignInAsync(
+                    A<HttpContext>._, A<string>._, A<ClaimsPrincipal>._, A<AuthenticationProperties>._))
+                .MustHaveHappened();
+        }
+
+        private void GivenSignInIsSuccessful(AdminUser expectedAdmin, List<DelegateUser> expectedDelegates)
```

Comment:
Could we move the private method at the end of the file (first public methods, then private)? 🙏 

Also, do we need to keep the expectedAdmin & expectedDelegates in the test or can they be part of the GivenSignInIsSuccessful (I guess it would need a rename to admin/delegates)